File: scripts/foundation/upgrade.py
#!/usr/bin/env python3
"""Offline, fail-closed source-snapshot rail. See docs/foundation-canary.md.

Release bundles are reviewed local code, not a remote command channel. Commands
are fixed here; plans cannot supply shell commands or writable destinations.
"""
from __future__ import annotations

import argparse
from contextlib import contextmanager
import fcntl
import hashlib
import json
import os
from pathlib import Path, PurePosixPath
import subprocess
import sys

SCHEMA = 1
RAIL = "sidebar-width-snapshot"
PAYLOAD = "src/foundation/sidebar-width.ts"
LOCK = "foundation.lock.json"
MANIFEST = "foundation.json"
ACTIONS = {"sidebar-finite-width": ["bun", "run", "test:foundation"]}
CHECKS = {
    "business": ["bun", "run", "test:business"],
    "types": ["bun", "run", "typecheck"],
    "build": ["bun", "run", "build", "--webpack"],
}
OWNERS = {"consumed", "vendored", "application", "generated"}
# These alone are omitted from source fingerprints. Ownership labels do NOT
# grant permission to hide source under a user-defined generated rule.
GENERATED = ["node_modules/", ".next/", "out/", ".turbo/", ".foundation/", "next-env.d.ts", "tsconfig.tsbuildinfo"]


class UpgradeError(Exception):
    pass


def require(condition: bool, message: str) -> None:
    if not condition:
        raise UpgradeError(message)


def digest(data: bytes) -> str:
    return hashlib.sha256(data).hexdigest()


def encode(value: object) -> bytes:
    return (json.dumps(value, sort_keys=True, indent=2) + "\n").encode()


def fingerprint(value: object) -> str:
    return digest(encode(value))


def safe_path(root: Path, relative: str) -> Path:
    require(isinstance(relative, str) and bool(relative), "Empty/invalid path")
    path = PurePosixPath(relative)
    require(not path.is_absolute() and ".." not in path.parts and "\\" not in relative,
            f"Unsafe path: {relative}")
    require(str(path) == relative.rstrip("/"), f"Noncanonical path: {relative}")
    current = root
    require(root.absolute() == root.resolve(), f"Symlink root or ancestor: {root}")
    for part in path.parts:
        current = current / part
        require(not current.is_symlink(), f"Symlink path: {relative}")
        if current.is_file():
            require(current.stat().st_nlink == 1, f"Hardlinked path: {relative}")
    return current


def read_json(path: Path) -> dict:
    try:
        value = json.loads(path.read_text())
    except (OSError, ValueError) as error:
        raise UpgradeError(f"Cannot read JSON {path}: {error}") from error
    require(isinstance(value, dict), f"Expected object: {path}")
    require(type(value.get("schemaVersion")) is int and value["schemaVersion"] == SCHEMA,
            f"Unsupported schema: {path}")
    return value


def atomic_json(path: Path, value: dict) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    temporary = path.with_name(path.name + ".tmp")
    require(not temporary.is_symlink(), f"Symlink temporary file: {temporary}")
    # Exclusive creation refuses stale files as well as symlink/hardlink aliases.
    with temporary.open("xb") as output:
        output.write(encode(value))
    os.replace(temporary, path)


def matches(path: str, rule: str) -> bool:
    return path.startswith(rule) if rule.endswith("/") else path == rule


def manifest(app: Path) -> dict:
    value = read_json(safe_path(app, MANIFEST))
    require(value.get("rail") == RAIL, "Unsupported rail")
    rules = value.get("ownership")
    require(isinstance(rules, dict) and bool(rules), "Missing ownership rules")
    for path, owner in rules.items():
        safe_path(app, path)
        require(owner in OWNERS, f"Unknown owner: {owner}")
    require(all(rules.get(path) == "generated" for path in GENERATED), "Generated paths are reserved")
    require(all(not any(matches(path, reserved) for reserved in GENERATED) or path in GENERATED
                for path in rules), "Cannot place source ownership inside generated paths")
    require(rules.get("src/foundation/") == "consumed", "Missing consumed boundary")
    require(all(owner != "consumed" or path == "src/foundation/" for path, owner in rules.items()),
            "Only src/foundation/ may be consumed on this rail")
    require(rules.get(LOCK) == "generated", "Lock must be generated")
    require(all(owner != "generated" or path in GENERATED + [LOCK] for path, owner in rules.items()),
            "Source cannot be hidden as generated")
    return value


def owner_of(value: dict, path: str) -> str:
    rules = [rule for rule in value["ownership"] if matches(path, rule)]
    require(bool(rules), f"Unclassified path: {path}")
    return value["ownership"][max(rules, key=len)]


def source_files(app: Path) -> dict[str, str]:
    value = manifest(app)
    result = {}
    for directory, dirs, files in os.walk(app, followlinks=False):
        parent = Path(directory)
        for name in list(dirs):
            path = (parent / name).relative_to(app).as_posix() + "/"
            if any(matches(path, rule) for rule in GENERATED):
                dirs.remove(name)
            else:
                safe_path(app, path)
        for name in files:
            path = (parent / name).relative_to(app).as_posix()
            if path == LOCK or any(matches(path, rule) for rule in GENERATED):
                continue
            file = safe_path(app, path)
            owner = owner_of(value, path)
            if owner != "consumed":
                result[path] = digest(file.read_bytes())
    return result


def installed(app: Path) -> dict[str, str]:
    root = safe_path(app, "src/foundation")
    require(root.is_dir(), "Missing consumed foundation")
    files = {}
    for path in root.rglob("*"):
        relative = path.relative_to(app).as_posix()
        safe_path(app, relative)
        if path.is_file():
            files[relative] = digest(path.read_bytes())
    return files


def catalogue(releases: Path) -> dict:
    value = read_json(safe_path(releases, "catalogue.json"))
    require(value.get("rail") == RAIL, "Catalogue rail mismatch")
    require(isinstance(value.get("releases"), dict) and bool(value["releases"]), "Missing releases")
    for version, release in value["releases"].items():
        require(isinstance(release, dict), "Invalid release")
        require(isinstance(release.get("files"), dict) and release["files"].keys() == {PAYLOAD},
                "Unsafe overwrite: unsupported release destinations")
        require(isinstance(release.get("from"), list) and
                all(isinstance(base, str) and base in value["releases"] for base in release["from"]),
                "Missing or unsupported baselines")
        require(isinstance(release.get("actions"), list) and
                all(isinstance(action, str) for action in release["actions"]),
                "Missing action/migration declaration")
        require(len(set(release["actions"])) == len(release["actions"]), "Duplicate action")
        require(all(action in ACTIONS for action in release["actions"]), "Unsupported action; upgrade the tooling")
        if release["from"]:
            require(bool(release["actions"]), "Missing action evidence requirement")
        for destination, sha in release["files"].items():
            payload = safe_path(releases, f"{version}/{destination}")
            require(payload.is_file() and digest(payload.read_bytes()) == sha, f"Corrupt release: {version}")
    require(value.get("latest") in value["releases"], "Unknown latest release")
    return value


def baseline(app: Path, catalog: dict) -> dict:
    lock = read_json(safe_path(app, LOCK))
    require(lock.get("rail") == RAIL, "Lock rail mismatch")
    require(lock.get("status") in {"baseline", "verified"}, "Incomplete upgrade; verify pending plan or restore from git")
    release = catalog["releases"].get(lock.get("release"))
    require(release is not None, "Unsupported baseline")
    require(lock.get("releaseDigest") == fingerprint(release), "Stale baseline release metadata")
    require(lock.get("files") == release["files"] == installed(app), "Consumed foundation drift; do not overwrite local edits")
    return lock


def discover(app: Path, releases: Path) -> dict:
    manifest(app)
    catalog = catalogue(releases)
    lock = baseline(app, catalog)
    targets = [version for version, release in catalog["releases"].items() if lock["release"] in release["from"]]
    return {"schemaVersion": SCHEMA, "rail": RAIL, "current": lock["release"],
            "availableTargets": targets, "latest": catalog["latest"]}


def plan(app: Path, releases: Path, target: str) -> dict:
    value = manifest(app)
    catalog = catalogue(releases)
    lock = baseline(app, catalog)
    require(target in catalog["releases"], "Unknown target")
    release = catalog["releases"][target]
    require(lock["release"] in release["from"], "Unsupported or stale baseline-to-target transition")
    for path in release["files"]:
        require(owner_of(value, path) == "consumed", f"Unsafe overwrite of {path}")
    result = {
        "schemaVersion": SCHEMA, "rail": RAIL, "from": lock["release"], "to": target,
        "catalogueDigest": fingerprint(catalog), "sourceDigest": fingerprint(source_files(app)),
        "toolDigest": digest(Path(__file__).read_bytes()),
        "lockDigest": fingerprint(lock),
        "changes": [{"path": path, "before": lock["files"][path], "after": sha}
                    for path, sha in sorted(release["files"].items())],
        "actions": [{"id": action, "command": ACTIONS[action]} for action in release["actions"]],
        "verification": [{"id": key, "command": command} for key, command in CHECKS.items()],
    }
    result["id"] = fingerprint(result)
    return result


def apply(app: Path, releases: Path, approved: dict) -> dict:
    require(approved == plan(app, releases, approved.get("to", "")), "Stale or modified upgrade plan")
    catalog = catalogue(releases)
    release = catalog["releases"][approved["to"]]
    # Pending is durable BEFORE touching consumed files. An interruption can never
    # look completed. Re-verify a fully copied tree, otherwise restore from git.
    pending = {"schemaVersion": SCHEMA, "rail": RAIL, "release": approved["to"],
               "releaseDigest": fingerprint(release), "files": release["files"],
               "status": "pending", "plan": approved}
    atomic_json(safe_path(app, LOCK), pending)
    for path in release["files"]:
        destination = safe_path(app, path)
        destination.write_bytes(safe_path(releases, f"{approved['to']}/{path}").read_bytes())
    return {"schemaVersion": SCHEMA, "status": "pending", "planId": approved["id"]}


def verification_context(app: Path, releases: Path, approved: dict) -> dict:
    catalog = catalogue(releases)
    lock = read_json(safe_path(app, LOCK))
    require(lock.get("status") in {"pending", "verified"} and lock.get("plan") == approved,
            "Missing applied plan; completion cannot be asserted")
    require(fingerprint(catalog) == approved.get("catalogueDigest"), "Catalogue changed since plan")
    require(digest(Path(__file__).read_bytes()) == approved.get("toolDigest"), "Upgrade tooling changed since plan")
    require(fingerprint(source_files(app)) == approved.get("sourceDigest"), "App-owned or vendored source changed since plan")
    release = catalog["releases"].get(approved.get("to"))
    require(release is not None and lock.get("release") == approved["to"] and
            lock.get("releaseDigest") == fingerprint(release) and
            lock.get("files") == release["files"] == installed(app), "Unverifiable installed target")
    require(approved.get("actions") == [{"id": action, "command": ACTIONS[action]} for action in release["actions"]],
            "Missing action/migration evidence requirement")
    require(approved.get("verification") == [{"id": key, "command": command} for key, command in CHECKS.items()],
            "Missing verification requirements")
    unsigned = {key: value for key, value in approved.items() if key != "id"}
    require(approved.get("id") == fingerprint(unsigned), "Invalid plan identity")
    return lock


def verify(app: Path, releases: Path, approved: dict) -> dict:
    lock = verification_context(app, releases, approved)
    # Re-verification invalidates any earlier success before a command runs.
    lock["status"] = "pending"
    lock.pop("evidenceDigest", None)
    atomic_json(safe_path(app, LOCK), lock)
    evidence = {"schemaVersion": SCHEMA, "rail": RAIL, "planId": approved["id"],
                "sourceDigest": approved["sourceDigest"], "release": approved["to"],
                "status": "failed", "results": []}
    evidence_path = safe_path(app, ".foundation/evidence.json")
    try:
        for check in approved["actions"] + approved["verification"]:
            if check["id"] == "build":
                # A successful no-op must not certify an old output as a new build.
                safe_path(app, ".next/server/app/dashboard.html").unlink(missing_ok=True)
            log = safe_path(app, f".foundation/{check['id']}.log")
            log.parent.mkdir(parents=True, exist_ok=True)
            with log.open("wb") as output:
                completed = subprocess.run(check["command"], cwd=app, stdout=output,
                                           stderr=subprocess.STDOUT, timeout=600, check=False,
                                           env={**os.environ, "NEXT_TELEMETRY_DISABLED": "1"})
            evidence["results"].append({**check, "exitCode": completed.returncode,
                                        "outputSha256": digest(log.read_bytes())})
            require(completed.returncode == 0, f"Verification failed: {check['id']}; see {log}")
        verification_context(app, releases, approved)
        artifact = safe_path(app, ".next/server/app/dashboard.html")
        require(artifact.is_file(), "Missing built dashboard artifact")
        evidence["artifacts"] = {".next/server/app/dashboard.html": digest(artifact.read_bytes())}
        evidence["status"] = "verified"
        atomic_json(evidence_path, evidence)
        lock["status"] = "verified"
        lock["evidenceDigest"] = fingerprint(evidence)
        atomic_json(safe_path(app, LOCK), lock)
        return evidence
    except (UpgradeError, OSError, subprocess.TimeoutExpired) as error:
        evidence["error"] = str(error)
        atomic_json(evidence_path, evidence)
        raise UpgradeError(str(error)) from error


def audit(app: Path, releases: Path) -> dict:
    """Check completion against this exact tree and retained executable evidence."""
    lock = read_json(safe_path(app, LOCK))
    require(lock.get("status") == "verified", "Completion is unverified")
    approved = lock.get("plan")
    require(isinstance(approved, dict), "Missing completion plan")
    verification_context(app, releases, approved)
    evidence = read_json(safe_path(app, ".foundation/evidence.json"))
    require(lock.get("evidenceDigest") == fingerprint(evidence) and evidence.get("status") == "verified" and
            evidence.get("rail") == RAIL and evidence.get("planId") == approved["id"] and
            evidence.get("sourceDigest") == approved["sourceDigest"] and evidence.get("release") == approved["to"],
            "Missing or altered completion evidence")
    artifacts = evidence.get("artifacts", {})
    require(isinstance(artifacts, dict) and artifacts.keys() == {".next/server/app/dashboard.html"},
            "Missing built artifact evidence")
    for path, sha in artifacts.items():
        artifact = safe_path(app, path)
        require(artifact.is_file() and digest(artifact.read_bytes()) == sha, "Built artifact changed or missing")
    expected = approved["actions"] + approved["verification"]
    results = evidence.get("results", [])
    require(len(results) == len(expected), "Missing action or verification evidence")
    for check, result in zip(expected, results):
        require(result.get("id") == check["id"] and result.get("command") == check["command"] and
                result.get("exitCode") == 0 and result.get("outputSha256") ==
                digest(safe_path(app, f".foundation/{check['id']}.log").read_bytes()),
                f"Unverifiable completion: {check['id']}")
    return evidence


@contextmanager
def exclusive(app: Path):
    guard = safe_path(app, ".foundation/upgrade.guard")
    guard.parent.mkdir(parents=True, exist_ok=True)
    with guard.open("a") as handle:
        try:
            fcntl.flock(handle, fcntl.LOCK_EX | fcntl.LOCK_NB)
        except BlockingIOError as error:
            raise UpgradeError("Another upgrade command is active") from error
        try:
            yield
        finally:
            fcntl.flock(handle, fcntl.LOCK_UN)


def execute(args: argparse.Namespace) -> dict:
    if args.command == "discover":
        return discover(args.app, args.releases)
    if args.command == "plan":
        require(bool(args.target), "plan requires --target")
        return plan(args.app, args.releases, args.target)
    if args.command == "audit":
        return audit(args.app, args.releases)
    require(args.plan is not None, f"{args.command} requires --plan")
    approved = read_json(args.plan)
    return (apply if args.command == "apply" else verify)(args.app, args.releases, approved)


def main() -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("command", choices=["discover", "plan", "apply", "verify", "audit"])
    parser.add_argument("--app", type=Path, default=Path("apps/demo"))
    parser.add_argument("--releases", type=Path, default=Path("foundation/releases"))
    parser.add_argument("--target")
    parser.add_argument("--plan", type=Path)
    args = parser.parse_args()
    try:
        with exclusive(args.app):
            result = execute(args)
        print(encode(result).decode(), end="")
        return 0
    except (UpgradeError, OSError, TypeError, KeyError, ValueError) as error:
        print(json.dumps({"schemaVersion": SCHEMA, "status": "blocked", "error": str(error)}), file=sys.stderr)
        return 1


if __name__ == "__main__":
    sys.exit(main())

File: platform/tooling/starter-upgrade/upgrade.ts
#!/usr/bin/env node
/** Local package upgrades. Release metadata cannot provide commands or write destinations. */
import * as fs from "node:fs";
import * as path from "node:path";
import { spawnSync } from "node:child_process";
import { pathToFileURL } from "node:url";
import { parseArgs } from "node:util";
import { validateConsumerImports } from "./consumer-boundary.ts";
import {
  SCHEMA, PACKAGE, BOUNDARY, PAYLOADS, LOCK, EVIDENCE, ARTIFACT,
  requireThat, digest, fingerprint, equal, encode, safePath, readJson, atomicJson,
  manifest, ownerOf, installed, sourceFiles,
  type Catalogue, type Lock, type Plan, type Check, type Evidence,
} from "./model.ts";
export * from "./model.ts";

export const ROOT = path.resolve(import.meta.dirname, "../../..");
export const RELEASES = path.join(ROOT, "apps/demo/qa/fixtures/starter-releases");
export const ACTIONS: Record<string, string[]> = { "sidebar-finite-width": ["bun", "run", "test:starter-policy"] };
export const CHECKS: Check[] = [
  { id: "business", command: ["bun", "run", "test:business"] },
  { id: "types", command: ["bun", "run", "typecheck"] },
  { id: "build", command: ["bun", "run", "build", "--webpack"] },
];
const EXPORTS = { ".": { types: "./dist/index.d.ts", default: "./dist/index.js" } };
export function packageContract(directory: string, version: string): void {
  const pkg = JSON.parse(fs.readFileSync(path.join(directory, "package.json"), "utf8")) as Record<string, unknown>;
  requireThat(pkg.name === PACKAGE && pkg.version === version && pkg.type === "module" && equal(pkg.exports, EXPORTS), "Invalid starter package identity or exports");
  requireThat(!pkg.dependencies && !pkg.devDependencies && !pkg.peerDependencies && !pkg.imports && !pkg.scripts, "Release package must be standalone and have no install scripts");
}
export function catalogue(releases: string): Catalogue {
  const value = readJson<Catalogue>(safePath(releases, "catalogue.json"));
  requireThat(value.package === PACKAGE && value.releases && typeof value.releases === "object" && !Array.isArray(value.releases), "Invalid starter catalogue");
  for (const [version, release] of Object.entries(value.releases)) {
    requireThat(/^\d+\.\d+\.\d+$/.test(version) && release && typeof release === "object", "Invalid release");
    requireThat(release.files && equal(Object.keys(release.files).sort(), [...PAYLOADS].sort()), "Unsafe overwrite: unsupported release destinations");
    requireThat(Array.isArray(release.from) && release.from.every(b => typeof b === "string" && Object.hasOwn(value.releases, b)), "Missing or unsupported baselines");
    requireThat(Array.isArray(release.actions) && release.actions.every(a => typeof a === "string" && Object.hasOwn(ACTIONS, a)), "Missing or unsupported action; upgrade the tooling");
    requireThat(new Set(release.actions).size === release.actions.length && (!release.from.length || release.actions.length), "Missing or duplicate action evidence requirement");
    requireThat(equal(release.affectedLayers, ["starter-sidebar-policy"]), "Missing or unknown affected layers");
    requireThat(["none", "low", "high", "critical"].includes(release.securityUrgency), "Missing or unknown security urgency");
    for (const [destination, sha] of Object.entries(release.files)) {
      const payload = safePath(releases, `${version}/${destination}`);
      requireThat(fs.existsSync(payload) && digest(fs.readFileSync(payload)) === sha, `Corrupt release: ${version}`);
    }
    requireThat(equal(installed(path.join(releases, version)), release.files), "Unexpected package artifact files");
    packageContract(path.join(releases, version, BOUNDARY), version);
  }
  requireThat(Object.hasOwn(value.releases, value.latest), "Unknown latest release");
  return value;
}
function toolDigest(): string {
  return fingerprint(["model.ts", "upgrade.ts", "consumer-boundary.ts"].map(f => digest(fs.readFileSync(path.join(import.meta.dirname, f)))));
}
export function baseline(app: string, catalog: Catalogue): Lock {
  const lock = readJson<Lock>(safePath(app, LOCK));
  requireThat(lock.package === PACKAGE, "Lock package mismatch");
  requireThat(["baseline", "verified"].includes(lock.status), "Incomplete upgrade; verify pending plan or restore from git");
  const release = catalog.releases[lock.release];
  requireThat(release, "Unsupported baseline");
  requireThat(lock.releaseDigest === fingerprint(release), "Stale baseline release metadata");
  requireThat(equal(lock.files, release.files) && equal(lock.files, installed(app)), "Consumed starter package drift; do not overwrite local edits");
  return lock;
}
export function discover(app: string, releases: string): object {
  sourceFiles(app);
  const catalog = catalogue(releases);
  const lock = baseline(app, catalog);
  const targets = Object.keys(catalog.releases).filter(v => catalog.releases[v].from.includes(lock.release));
  return { schemaVersion: SCHEMA, package: PACKAGE, current: lock.release, latest: catalog.latest, availableTargets: targets,
    targetDetails: Object.fromEntries(targets.map(v => [v, { affectedLayers: catalog.releases[v].affectedLayers, securityUrgency: catalog.releases[v].securityUrgency }])) };
}
export function plan(app: string, releases: string, target: string): Plan {
  const value = manifest(app), catalog = catalogue(releases), lock = baseline(app, catalog);
  requireThat(Object.hasOwn(catalog.releases, target), "Unknown target");
  const release = catalog.releases[target];
  requireThat(release.from.includes(lock.release), "Unsupported or stale baseline-to-target transition");
  for (const file of Object.keys(release.files)) requireThat(ownerOf(value, file) === "consumed", `Unsafe overwrite of ${file}`);
  const unsigned = {
    schemaVersion: SCHEMA, package: PACKAGE, from: lock.release, to: target,
    affectedLayers: release.affectedLayers, securityUrgency: release.securityUrgency,
    catalogueDigest: fingerprint(catalog), sourceDigest: fingerprint(sourceFiles(app)),
    toolDigest: toolDigest(), lockDigest: fingerprint(lock),
    changes: Object.entries(release.files).sort().map(([p, sha]) => ({ path: p, before: lock.files[p], after: sha })),
    actions: release.actions.map(id => ({ id, command: ACTIONS[id] })), verification: CHECKS,
  };
  return { ...unsigned, id: fingerprint(unsigned) };
}
export function apply(app: string, releases: string, approved: Plan): object {
  requireThat(equal(approved, plan(app, releases, approved.to)), "Stale or modified upgrade plan");
  const release = catalogue(releases).releases[approved.to];
  const lock: Lock = { schemaVersion: SCHEMA, package: PACKAGE, release: approved.to, releaseDigest: fingerprint(release), files: release.files, status: "pending", plan: approved };
  atomicJson(safePath(app, LOCK), lock);
  for (const file of Object.keys(release.files)) {
    const destination = safePath(app, file, true);
    fs.writeFileSync(destination + ".tmp", fs.readFileSync(safePath(releases, `${approved.to}/${file}`)), { flag: "wx" });
    fs.renameSync(destination + ".tmp", destination);
  }
  return { schemaVersion: SCHEMA, status: "pending", planId: approved.id };
}
export function verificationContext(app: string, releases: string, approved: Plan): Lock {
  const catalog = catalogue(releases), lock = readJson<Lock>(safePath(app, LOCK));
  requireThat(["pending", "verified"].includes(lock.status) && equal(lock.plan, approved), "Missing applied plan; completion cannot be asserted");
  requireThat(fingerprint(catalog) === approved.catalogueDigest, "Catalogue changed since plan");
  requireThat(toolDigest() === approved.toolDigest, "Upgrade tooling changed since plan");
  requireThat(fingerprint(sourceFiles(app)) === approved.sourceDigest, "Application source changed since plan");
  const release = catalog.releases[approved.to];
  requireThat(release && lock.package === PACKAGE && lock.release === approved.to && lock.releaseDigest === fingerprint(release) && equal(lock.files, release.files) && equal(lock.files, installed(app)), "Unverifiable installed target");
  requireThat(equal(approved.actions, release.actions.map(id => ({ id, command: ACTIONS[id] }))), "Missing action evidence requirement");
  requireThat(equal(approved.verification, CHECKS), "Missing verification requirements");
  const { id, ...unsigned } = approved;
  requireThat(id === fingerprint(unsigned), "Invalid plan identity");
  return lock;
}
/** The consumer must resolve the approved local package, never current workspace source. */
export function consumerBoundary(app: string): void {
  const dependency = path.join(app, "node_modules", PACKAGE);
  requireThat(fs.existsSync(dependency), "Starter dependency is not installed");
  const resolved = fs.realpathSync(dependency);
  const local = fs.realpathSync(safePath(app, BOUNDARY));
  requireThat(resolved === local || resolved.startsWith(path.join(ROOT, "node_modules") + path.sep) || resolved.startsWith(path.join(app, "node_modules") + path.sep), "Starter dependency resolves to workspace source instead of the consumed package");
  for (const file of PAYLOADS) {
    const expected = fs.readFileSync(safePath(app, file, true));
    requireThat(digest(fs.readFileSync(path.join(resolved, file.slice(BOUNDARY.length)))) === digest(expected), "Installed dependency differs from the consumed starter package; reinstall local dependencies");
  }
  validateConsumerImports(app, ROOT);
}
export type Runner = (app: string, check: Check, log: string) => number;
export const runCheck: Runner = (app, check, log) => {
  const fd = fs.openSync(log, "w");
  try {
    const result = spawnSync(check.command[0], check.command.slice(1), { cwd: app, stdio: ["ignore", fd, fd], timeout: 600_000, env: { ...process.env, NEXT_TELEMETRY_DISABLED: "1" } });
    if (result.error) throw result.error;
    return result.status ?? 1;
  } finally { fs.closeSync(fd); }
};
export function verify(app: string, releases: string, approved: Plan, runner: Runner = runCheck): Evidence {
  const lock = verificationContext(app, releases, approved);
  lock.status = "pending"; delete lock.evidenceDigest;
  atomicJson(safePath(app, LOCK), lock);
  const evidence: Evidence = { schemaVersion: SCHEMA, package: PACKAGE, planId: approved.id, sourceDigest: approved.sourceDigest, release: approved.to, status: "failed", results: [] };
  const evidencePath = safePath(app, EVIDENCE);
  fs.mkdirSync(path.dirname(evidencePath), { recursive: true });
  try {
    consumerBoundary(app);
    for (const check of [...approved.actions, ...approved.verification]) {
      if (check.id === "build") fs.rmSync(safePath(app, ARTIFACT), { force: true });
      const log = safePath(app, `.starter-upgrade/${check.id}.log`);
      const exitCode = runner(app, check, log);
      evidence.results.push({ ...check, exitCode, outputSha256: digest(fs.readFileSync(log)) });
      requireThat(exitCode === 0, `Verification failed: ${check.id}; see ${log}`);
    }
    verificationContext(app, releases, approved); consumerBoundary(app);
    const artifact = safePath(app, ARTIFACT);
    requireThat(fs.existsSync(artifact), "Missing built dashboard artifact");
    evidence.artifacts = { [ARTIFACT]: digest(fs.readFileSync(artifact)) };
    evidence.status = "verified"; atomicJson(evidencePath, evidence);
    lock.status = "verified"; lock.evidenceDigest = fingerprint(evidence); atomicJson(safePath(app, LOCK), lock);
    return evidence;
  } catch (error) {
    evidence.error = String(error); atomicJson(evidencePath, evidence); throw error;
  }
}
export function audit(app: string, releases: string): Evidence {
  const lock = readJson<Lock>(safePath(app, LOCK));
  requireThat(lock.status === "verified", "Completion is unverified");
  requireThat(lock.plan, "Missing completion plan");
  const approved = lock.plan;
  verificationContext(app, releases, approved); consumerBoundary(app);
  const evidence = readJson<Evidence>(safePath(app, EVIDENCE));
  requireThat(lock.evidenceDigest === fingerprint(evidence) && evidence.status === "verified" && evidence.package === PACKAGE && evidence.planId === approved.id && evidence.sourceDigest === approved.sourceDigest && evidence.release === approved.to, "Missing or altered completion evidence");
  requireThat(evidence.artifacts && equal(Object.keys(evidence.artifacts), [ARTIFACT]), "Missing built artifact evidence");
  for (const [p, sha] of Object.entries(evidence.artifacts)) requireThat(digest(fs.readFileSync(safePath(app, p))) === sha, "Built artifact changed or missing");
  const expected = [...approved.actions, ...approved.verification];
  requireThat(Array.isArray(evidence.results) && evidence.results.length === expected.length, "Missing action or verification evidence");
  expected.forEach((check, i) => {
    const result = evidence.results[i];
    requireThat(result.id === check.id && equal(result.command, check.command) && result.exitCode === 0 && result.outputSha256 === digest(fs.readFileSync(safePath(app, `.starter-upgrade/${check.id}.log`))), `Unverifiable completion: ${check.id}`);
  });
  return evidence;
}
export function exclusive<T>(app: string, work: () => T): T {
  const guard = safePath(app, ".starter-upgrade/upgrade.guard");
  fs.mkdirSync(path.dirname(guard), { recursive: true });
  try { fs.mkdirSync(guard); } catch { throw new Error("Another upgrade command is active or an interrupted command left upgrade.guard; inspect before removing it"); }
  try { return work(); } finally { fs.rmdirSync(guard); }
}
export function main(): void {
  try {
    const args = parseArgs({ allowPositionals: true, options: { app: { type: "string", default: "apps/demo" }, releases: { type: "string", default: RELEASES }, target: { type: "string" }, plan: { type: "string" } } });
    const [command] = args.positionals;
    const app = path.resolve(args.values.app), releases = path.resolve(args.values.releases);
    const result = exclusive(app, () => {
      if (command === "discover") return discover(app, releases);
      if (command === "plan") { requireThat(args.values.target, "plan requires --target"); return plan(app, releases, args.values.target); }
      if (command === "audit") return audit(app, releases);
      requireThat(command === "apply" || command === "verify", "Expected discover, plan, apply, verify or audit");
      requireThat(args.values.plan, `${command} requires --plan`);
      const approved = readJson<Plan>(args.values.plan);
      return command === "apply" ? apply(app, releases, approved) : verify(app, releases, approved);
    });
    process.stdout.write(encode(result));
  } catch (error) { process.stderr.write(encode({ schemaVersion: SCHEMA, status: "blocked", error: String(error) })); process.exitCode = 1; }
}
if (process.argv[1] && pathToFileURL(path.resolve(process.argv[1])).href === import.meta.url) main();
